src/main/scala/org/ergoplatform/nodeView/history/ErgoHistory.scala

```diff
@@ -258,21 +260,40 @@ object ErgoHistory extends ScorexLogging {
 
   // check if there is possible database corruption when there is header after
   // recognized blockchain tip marked as invalid
-  protected[nodeView] def repairIfNeeded(history: ErgoHistory): Boolean = history.historyStorage.synchronized {
+  protected[nodeView] def repairIfNeeded(history: ErgoHistory): Unit = history.historyStorage.synchronized {
+    val RepairDepth = 128
+
     val bestHeaderHeight = history.headersHeight
-    val bestFullBlockHeight = history.bestFullBlockOpt.map(_.height).getOrElse(-1)
-    val afterHeaders = history.headerIdsAtHeight(bestHeaderHeight + 1)
 
-    if (bestHeaderHeight == bestFullBlockHeight && afterHeaders.nonEmpty) {
-      log.warn("Found suspicious continuation, clearing it...")
-      afterHeaders.map { hId =>
-        history.forgetHeader(hId)
+    @tailrec
+    def checkHeightsFrom(h: Int): Unit = {
+      val headerIds = history.headerIdsAtHeight(h)
+      if (headerIds.nonEmpty) {
+        val notInvalidHeaders = headerIds.filter { headerId =>
+          if (history.isSemanticallyValid(headerId) == Invalid) {
+            log.warn(s"Clearing invalid header: $headerId at height $h")
+            history.forgetHeader(headerId)
+            false
+          } else {
+            true
+          }
+        }
+        val updatedHeightIdsValue: Array[Byte] = notInvalidHeaders.foldLeft(Array.empty[Byte]) { case (acc, id) =>
+          acc ++ idToBytes(id)
+        }
+        if(updatedHeightIdsValue.isEmpty) {
+          //could be the case after bestHeaderHeight
+          history.historyStorage.remove(Array(history.heightIdsKey(h)), Nil)
```

Comment:
Shouldn't we also remove the Modifier at this height? Not just index? Or is there an idempotency in place that the modifer gets overriden if index is empty?